ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java

```diff
@@ -72,18 +71,16 @@ public static boolean is_active_validator(Validator validator, UnsignedLong epoc
    *     <a>https://github.com/ethereum/eth2.0-specs/blob/v0.8.0/specs/core/0_beacon-chain.md#get_active_validator_indices</a>
    */
   public static List<Integer> get_active_validator_indices(BeaconState state, UnsignedLong epoch) {
-    List<Integer> active_validator_indices = Collections.synchronizedList(new ArrayList<>());
     List<Validator> validators = state.getValidators();
-    IntStream.range(0, validators.size())
+    return IntStream.range(0, validators.size())
         .parallel()
-        .forEachOrdered(
-            index -> {
-              if (is_active_validator(validators.get(index), epoch)) {
-                active_validator_indices.add(index);
-              }
-            });
-
-    return active_validator_indices;
+        .flatMap(
```

Comment:
Does flatMap sustain the order of elements here? I believe we had an issue with a similar change in the past due to not preserving the order of validator indices.

Reply:
@cemozerr Good question. I didn't find any mentions that `flatMap` makes a stream unordered. My minitest also confirms this. However this is not relevant any more, since I change this method in the next commit

Reply:
flatMap preserves ordering but making a stream parallel may not (you can have parallel, ordered streams but that restricts how much parallelisation can happen). I can't recall how or if you can choose between ordered or unordered parallel streams so tend to just assume parallel streams won't preserve order unless I go check more carefully.

Reply:
From my understanding of the spec if the original stream is `ordered` and all pipeline operators preserve ordering (unlike e.g. `.unordered()` operator) then collector would make ordered list.
If e.g. `is_active_validator` is heavy operation (which is actually not the case), then parallelization would yield good gain here. 

E.g. the following test:
```java
    IntStream.range(0, 10)
        .parallel()
        .flatMap(i -> {
          LOG.info("Calculating " + i);
          Thread.sleep(1000);
          return i % 2 == 0 ? IntStream.of(i) : IntStream.empty();
        })
        .forEachOrdered(i -> LOG.info(i));
```
Would print: 
```log
11:59:25.985+03:00 | ForkJoinPool.commonPool-worker-7  | Calculating 1
11:59:25.985+03:00 | ForkJoinPool.commonPool-worker-15 | Calculating 7
11:59:25.985+03:00 | ForkJoinPool.commonPool-worker-9  | Calculating 5
11:59:25.985+03:00 | ForkJoinPool.commonPool-worker-5  | Calculating 2
11:59:25.985+03:00 | ForkJoinPool.commonPool-worker-3  | Calculating 8
11:59:25.985+03:00 | main                              | Calculating 6
11:59:25.985+03:00 | ForkJoinPool.commonPool-worker-11 | Calculating 0
11:59:25.985+03:00 | ForkJoinPool.commonPool-worker-13 | Calculating 9
11:59:26.994+03:00 | ForkJoinPool.commonPool-worker-7  | Calculating 3
11:59:26.994+03:00 | ForkJoinPool.commonPool-worker-13 | Calculating 4
11:59:26.995+03:00 | ForkJoinPool.commonPool-worker-11 | 0
11:59:26.996+03:00 | ForkJoinPool.commonPool-worker-11 | 2
11:59:27.995+03:00 | ForkJoinPool.commonPool-worker-13 | 4
11:59:27.995+03:00 | ForkJoinPool.commonPool-worker-13 | 6
11:59:27.995+03:00 | ForkJoinPool.commonPool-worker-13 | 8
```